### PR TITLE
feat(websocket): Allow wss connections on IP addresses

### PR DESCRIPTION
**Pull request by oblique**
## Notes & open questions

It is completely valid to establish TLS connection with just an IP if:

1. You add your self-sign certificate via [`add_trust`](https://docs.rs/libp2p/latest/libp2p/websocket/tls/struct.Builder.html#method.add_trust).
2. You create a valid certificate for an IP. For example 1.1.1.1 has a valid certificate.

Example with rustls:

```rust
use rustls::{ClientConfig, ClientConnection, RootCertStore, Stream};
use std::io::{Read, Write};
use std::net::TcpStream;
use std::sync::Arc;
use webpki_roots::TLS_SERVER_ROOTS;

fn main() {
    let provider = rustls::crypto::ring::default_provider();
    let cert_store = TLS_SERVER_ROOTS.iter().cloned().collect::<RootCertStore>();

    let client = ClientConfig::builder_with_provider(provider.into())
        .with_safe_default_protocol_versions()
        .unwrap()
        .with_root_certificates(cert_store)
        .with_no_client_auth();

    let server_name = "1.1.1.1".try_into().unwrap();
    let mut conn = ClientConnection::new(Arc::new(client), server_name).unwrap();
    let mut sock = TcpStream::connect("1.1.1.1:443").unwrap();

    let mut io = Stream::new(&mut conn, &mut sock);

    io.write_all(b"GET / HTTP/1.1\r\n\r\n").unwrap();

    let mut resp = Vec::new();
    io.read_to_end(&mut resp).unwrap();

    println!("{}", String::from_utf8(resp).unwrap());
}
```

## Change checklist

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: transports/websocket/src/framed.rs [hunk 1]
```diff
@@ -478,19 +477,21 @@ fn parse_ws_dial_addr<T>(addr: Multiaddr) -> Result<WsAddress, Error<T>> {
...
         match (ip, tcp) {
             (Some(Protocol::Ip4(ip)), Some(Protocol::Tcp(port))) => {
-                break (format!("{ip}:{port}"), None)
+                let server_name = ServerName::IpAddress(IpAddr::V4(ip).into());
+                break (format!("{ip}:{port}"), server_name);
             }
             (Some(Protocol::Ip6(ip)), Some(Protocol::Tcp(port))) => {
-                break (format!("{ip}:{port}"), None)
+                let server_name = ServerName::IpAddress(IpAddr::V6(ip).into());
+                break (format!("[{ip}]:{port}"), server_name);
```
Comment (oblique) [inline transports/websocket/src/framed.rs:478]: I noticed that `[]` were missing around the IPv6. Is it ok to have it in this PR? Or should I open a separated one?

File: transports/websocket/src/framed.rs [hunk 2]
```diff
@@ -468,19 +467,21 @@ fn parse_ws_dial_addr<T>(addr: Multiaddr) -> Result<WsAddress, Error<T>> {
     let mut protocols = addr.iter();
     let mut ip = protocols.next();
     let mut tcp = protocols.next();
-    let (host_port, dns_name) = loop {
+    let (host_port, server_name) = loop {
         match (ip, tcp) {
             (Some(Protocol::Ip4(ip)), Some(Protocol::Tcp(port))) => {
-                break (format!("{ip}:{port}"), None)
+                let server_name = ServerName::IpAddress(IpAddr::V4(ip).into());
+                break (format!("{ip}:{port}"), server_name);
```
Comment (jxs) [inline transports/websocket/src/framed.rs:474]:
```suggestion
                break (format!("[{ip}]:{port}"), server_name);
```

File: transports/websocket/src/framed.rs (hunk 1 above)
Comment (jxs) [reply transports/websocket/src/framed.rs:478]: yeah, let's do it in this one, I just suggested it

**Review by jxs — APPROVED** [review]
LGTM Yiannis

File: transports/websocket/src/framed.rs (hunk 2 above)
Comment (oblique) [reply transports/websocket/src/framed.rs:474]: IPv4 addresses do not need `[]`

**Pull request merged**